contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -1180,6 +1191,133 @@ tle_name_comparison(const char *tlename, const char *identifier)
 		return (0 == strcmp(tlename, identifier));
 }
 
+/*
+* A special case of the get_object_address_relobject() function, specifically
+* for the case of triggers in tsql dialect. We add a pg_trigger lookup to search
+* for the relation that the trigger is associated with, since the relation name
+* is not supplied by the user, and thus not a part of the *object list.
+*/
+static ObjectAddress
+get_trigger_object_address(List *object, Relation *relp, bool missing_ok)
+{
+	ObjectAddress address;
+	Relation	relation = NULL;
+	const char *depname;
+	Oid			reloid;
+	Relation	tgrel;
+	ScanKeyData		key;
+	SysScanDesc tgscan;
+	HeapTuple	tuple;
+	const char	*trigger_physical_schema = NULL;
+	const char	*trigger_logical_schema = NULL;
+	const char	*cur_dbo_physical_schema= NULL;
+	bool 	found_trigger = false;
+	const char	*pg_trigger_physical_schema = NULL;
+	const char	*pg_trigger_logical_schema = NULL;
+	const char	*cur_physical_schema = NULL;
+
+	if (sql_dialect != SQL_DIALECT_TSQL)
+	{
+		address.classId = InvalidOid;
+		address.objectId = InvalidOid;
+		address.objectSubId = InvalidAttrNumber;
+		return address;
+	}
+	/* Extract name of dependent object. */
+	depname = strVal(llast(object));
+	if (list_length(object) > 1){
+		trigger_physical_schema = ((Value *)list_nth(object,0))->val.str;
+		trigger_logical_schema = get_logical_schema_name(trigger_physical_schema, true);
+	}
+
+	cur_dbo_physical_schema = get_dbo_schema_name(get_cur_db_name());
+
+	if (prev_get_trigger_object_address_hook)
+		return (*prev_get_trigger_object_address_hook)(object,relp,missing_ok);
+
+	/* 
+	* Get the table name of the trigger from pg_trigger. We know that
+	* trigger names are forced to be unique in the tsql dialect, so we
+	* can rely on searching for trigger name and schema name to find 
+	* the corresponding relation name.
+	*/
+	tgrel = table_open(TriggerRelationId, AccessShareLock);
+	ScanKeyInit(&key,
+					Anum_pg_trigger_tgname,
+					BTEqualStrategyNumber, F_NAMEEQ,
+					CStringGetDatum(depname));
+
+	tgscan = systable_beginscan(tgrel, TriggerRelidNameIndexId, false,
+									NULL, 1, &key);
+	while (HeapTupleIsValid(tuple = systable_getnext(tgscan)))
+	{
+		Form_pg_trigger pg_trigger = (Form_pg_trigger) GETSTRUCT(tuple);
+		if(!OidIsValid(pg_trigger->tgrelid))
+		{
+			reloid = InvalidOid;
+			break;
+		}
+		pg_trigger_physical_schema = get_namespace_name(get_rel_namespace(pg_trigger->tgrelid));
+		pg_trigger_logical_schema = get_logical_schema_name(pg_trigger_physical_schema, true);
+		cur_physical_schema = get_physical_schema_name(get_cur_db_name(),trigger_logical_schema);
+		if(namestrcmp(&(pg_trigger->tgname), depname) == 0){
+			reloid = pg_trigger->tgrelid;
+			relation = RelationIdGetRelation(reloid);
+			if (list_length(object) == 1 && 
+				strcasecmp(pg_trigger_physical_schema,cur_dbo_physical_schema) == 0)
+			{	
+				found_trigger = true;
+				RelationClose(relation);
```

Comment:
`RelationClose()` will be called twice for the same relation, once here and once on line 1281 - this might lead to unexpected behavior. Remove the `RelationClose()` inside the if() statements and keep the one on line 1281 so we don't decrement the reference count twice.